src/io/unix/sync.rs

```diff
@@ -59,39 +59,61 @@ impl File {
         self.size.load(Ordering::SeqCst)
     }
 
-    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
-        let offset = self.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
-        self.write_all_at(offset, buf).await
-    }
-
-    pub(crate) async fn write_append_all_buffers(
+    pub(crate) async fn write_append_writable_data<R: Send + 'static>(
         &self,
-        first_buf: Bytes,
-        second_buf: Bytes,
-    ) -> IOResult<()> {
-        let first_len = first_buf.len() as u64;
-        let second_len = second_buf.len() as u64;
-        let len = first_len + second_len;
-        let mut offset = self.size.fetch_add(len, Ordering::SeqCst);
+        c: impl WritableDataCreator<R>,
+    ) -> Result<R> {
+        let len = c.len();
+        let size = self.size.clone();
         let file = self.no_lock_fd.clone();
         if Self::can_run_inplace(len) {
             Self::inplace_sync_call(move || {
-                file.write_all_at(&first_buf, offset)?;
-                offset = offset + first_len;
-                file.write_all_at(&second_buf, offset)?;
-                Ok(())
+                let offset = size.fetch_add(len, Ordering::SeqCst);
+                let (res, data) = c.create(offset)?;
+                Self::write_data(file.as_ref(), offset, res)?;
+                Ok(data)
             })
         } else {
             Self::background_sync_call(move || {
-                file.write_all_at(&first_buf, offset)?;
-                offset = offset + first_len;
-                file.write_all_at(&second_buf, offset)?;
-                Ok(())
+                let offset = size.fetch_add(len, Ordering::SeqCst);
+                let (res, data) = c.create(offset)?;
+                Self::write_data(file.as_ref(), offset, res)?;
+                Ok(data)
             })
             .await
         }
     }
 
+    fn write_data(file: &StdFile, mut offset: u64, writable_data: WritableData) -> Result<()> {
+        let file_result: IOResult<()> = match writable_data {
+            WritableData::Single(bytes) => file.write_all_at(&bytes, offset),
+            WritableData::Double(b1, b2) => file.write_all_at(&b1, offset).and_then(|_| {
+                offset = offset + b1.len() as u64;
+                file.write_all_at(&b2, offset)
+            }),
+        };
+        file_result.map_err(|e| match e.kind() {
```

Comment:
I think, it is better to return `IoResult<()>` from all the functions within `File` (like it was before) and this error processing should be on `PartiallySerializedRecord` level (again: like it was before)